Updated traffic management concept guide (#5014)
istio/istio.io: content/en/docs/concepts/traffic-management/index.md

    - /docs/concepts/traffic-management/pilot.html
---

Istio’s traffic routing rules let you easily control the flow
of traffic and API calls between services. Istio simplifies configuration of
service-level properties like circuit breakers, timeouts, and retries, and makes
it easy to set up important tasks like A/B testing, canary rollouts, and staged
rollouts with percentage-based traffic splits. It also provides out-of-box
failure recovery features, which, combined with Istio’s
[observability](/docs/concepts/observability/) features, helps make your application

> more robust against failures of dependent services or the network.

Reviewer: I find this paragraph a little confusing.  You talk about the traffic management freatures in detail, then mention one of the core value propositions of Istio (observability) while leaving out the other major core value proposition of Istio (authenticated encrypted service communication).  I'm not sure what the best way to handle this problem in this paragraph.  A couple of options that come to mind are either - mention security, or remove the mention of observability.

Author: I think taking observability out is probably cleaner.

> Istio’s traffic management model relies on the {{< gloss >}}Envoy{{</ gloss >}}

Reviewer: `along with your services` is a little vague, but I'm not sure how to add enough detail without blowing people up with too much detail.

Author: "as sidecars to your services"?

proxies that are deployed along with your services. All traffic that your mesh

> services send and receive ({{< gloss >}}data plane{{</ gloss >}} traffic) is proxied through Envoy, making

Reviewer: parenthesis in the docs is generally discouraged.  Perhaps a better approach would be to rework this sentence a bit such as `All data-plane traffic within your service mesh is proxied through Envoy. This proxy model makes it easy for you to direct and control traffic within your mesh without making any changes to your services.`

this simplifies the sentence a bit and makes it less complex to understand.

Author: Will what "data plane traffic" means be obvious to all readers? The main reason for the parenthesized bit is to convey "this traffic is also referred to as data plane traffic" (though I *have* just added data plane to the glossary, so that might mitigate)

it easy to direct and control traffic around your mesh without making any
changes to your services.

If you’re interested in the details of how the features described in this guide
work, you can find out more about Istio’s traffic management architecture in the

> [Architecture](#architecture) section at the end of this document. The rest of

Reviewer: This wording is a little clunky for a link.  How about something like "If you are interested in the details of traffic management, please reference the architecture(link).

Author: I think that loses a little of the meaning - it's less the details of traffic management, more the details of how traffic management *works* (ie the implementation part). How about:

If you're interested in the details of how Istio's traffic management works, you can find out more  in the [Architecture]() section.

Reviewer 2: Don't lose the emphasis in the sentence that this guide is focused on "features", not architecture.

Author: How about:

If you're interested in the details of how the features described in this guide work, you can find out more in....

Reviewer 2: Yes, remove the "about Istio’s traffic management architecture" part. That works.

this guide introduces Istio’s traffic management features.

## Introducing Istio Traffic Management

> In order to direct traffic around your mesh, Istio needs to know where all your

Reviewer: around/withiin

Author: done

> endpoints are, and which services they belong to. To populate its own

Reviewer: its -> Istio's (mention istio as much as possible - part of programming the reader ;)

Author: It makes the sentence read very oddly though - the "it" isn't in any way ambiguous, and otherwise we're saying "Istio populates Istio's service registry", which sounds peculiar.

{{< gloss >}}service registry{{</ gloss >}}, Istio connects to a service

> discovery system. For example, if you've installed Istio on a Kubernetes cluster,

Reviewer: remove For example, and start the sentence with If.

Installed Istio on Kubernetes (remove a).

Author: I think we should leave in "For example", otherwise (per Srinath's concerns above) it's not obvious that this is just one of the platforms where you could install Istio. Also removing "cluster" makes the sentence a bit odd because we're talking specifically about a cluster: "the services and endpoints on that cluster".

Reviewer 2: Don't change this. P.S. you mean Shriram (no Srinath).

Author: Gah, of course I do. Sorry Shriram!

> then by default Istio knows about all the services and endpoints in that cluster.

Reviewer: then by default Istio knows -> then Istio automatically detects

Author: done

Using this service registry, the Envoy proxies can then direct traffic to the
relevant services. Most microservice-based applications have multiple instances
of each service workload to handle service traffic, sometimes referred to as a

> load balancing pool. By default, the Envoy proxies distribute traffic across

Reviewer: probably need a glossary for load balancing pool.  This is the first time I've really heard that terminology used.

Author: Is it not a fairly generic networking term? I don't think we use it in a particularly Istio-specific way, and adding glossary entries for terms that don't have Istio-specific meanings doesn't seem like a direction we should be going (also might confuse readers as they might think we *are* using it in an unusual way).

each service’s load balancing pool using a round-robin model, where requests are

> sent to each pool member in turn, returning to the top of the pool once each

Reviewer: something here feels technically off, although I'm not sure what it is.  Perhaps this paragraph is too much detail for the concepts section.  Any other reviewers have actionable suggestions for rework here?

Author: I think it's useful info because it explains what happens if you just use Istio out of the box and don't do any traffic configuration at all - your data plane traffic still gets sent to workload instances, it's just that you can't do anything more complicated than "send to that service w round robin load balancing between all its instances". We then excite and delight the users in the next paragraph with all the lovely things they can do if they use our traffic management resources. :)

Author: @frankbu, thoughts?

Reviewer 2: It's good as is.

service instance has received a request.

> While this basic service discovery and load balancing gives you a working

Reviewer: this->Istio's

Author: done

service mesh, it’s far from all that Istio can do. In many cases you might want

> more fine-grained control over what happens to your mesh traffic. For example,

Reviewer: remove For example please.

Author: done.

you might want to direct a particular percentage of traffic to a new version of
a service as part of A/B testing, or apply a different load balancing policy to

> traffic for a particular subset of service instances. You might also want to

Reviewer: for further followup work, a link to each of the tasks here would be absolutely awesome!!

Author: Will note that for a next iteration.

apply special rules to traffic coming into or out of your mesh, or add an
external dependency of your mesh to the service registry. You can do all this
and more by adding your own traffic configuration to Istio using Istio’s traffic
management API.

Like other Istio configuration, the API is implemented as Kubernetes custom
resources (CRDs), which you can configure using YAML, as you’ll see in the
examples.

> The rest of this guide looks at each of the traffic management API’s resources

Reviewer: looks at->examines

Author: done.

> The rest of this guide looks at each of the traffic management API’s resources

Reviewer: API's -> API

Author: done.

and what you can do with them. These resources are:

-   [Virtual services](#virtual-services)
-   [Destination rules](#destination-rules)
-   [Gateways](#gateways)
-   [Service entries](#service-entries)

> -   [Sidecars](#sidecars)

Reviewer: bangin work!  - I'd hiighly recommend not adding extra spaces though with bullets.  Just use `- [id](link)`.  Thew infrastructure sometimess is confused by extra spaces in md formatting.

Author: done. It's rendering fine but I do not wish to anger the Blackfriday parser....

This guide also gives an overview of some of the
[network resilience and testing features](#network-resilience-and-testing) that

> are built in to the API’s resources.

Reviewer: API's -> API.  We have one API :)

Author: Fair point. :)

## Virtual services {#virtual-services}

[Virtual services](/docs/reference/config/networking/v1alpha3/virtual-service/#VirtualService),
along with [destination rules](#destination-rules), are the key building blocks of Istio’s traffic
routing functionality. A virtual service lets you configure how requests are
routed to a service within an Istio service mesh, building on the basic
connectivity and discovery provided by Istio and your platform. Each virtual
service consists of a set of routing rules that are evaluated in order, letting
Istio match each given request to the virtual service to a specific real
destination within the mesh. Your mesh can require multiple virtual services or
none depending on your use case.

### Why use virtual services? {#why-use-virtual-services}

Virtual services play a key role in making Istio’s traffic management flexible

> and powerful. They do this by strongly decoupling where clients send their

Reviewer: this sentence is a lititle hard to follow.

Author: I'm not sure how to clarify further:

"...by strongly decoupling where clients send their requests from the destination workloads that actually implement those requests"?

Or we could combine the first two sentences but then we get a very long sentence, which is IMO even harder to follow.

Virtual services play a key role in making Istio's traffic management flexible and powerful by strongly decoupling where ....

requests from the destination workloads that actually implement them. Virtual
services also provide a rich way of specifying different traffic routing rules
for sending traffic to those workloads.

> So why is this so useful? Without virtual services, Envoy just distributes

Reviewer: Remove So? :)

Reviewer: remove just.

Author: done.

traffic using round-robin load balancing between all service instances, as

> described in the introduction. You can improve this behavior with what you know

Reviewer: remove as describe in the introduction.

Author: I think here it's useful to remind the user that "we talked about this before" (this is not entirely new info if you read from the top but that's where the details are), and gives someone who's just come into this section without reading the intro a pointer to where they can find out more.

> about the workloads: for example, some might represent a different version. This

Reviewer: remove : and replace with .

Reviewer: remove for example

Author: We're losing a relationship between the clauses if we break the sentences *and* remove "for example". "representing a different version" is an example of "something you might know about the workloads".

Could break into:
You can improve this behavior with what you know about the workloads. For example, some might represent a different version.

can be useful in A/B testing where you might want to configure traffic routes

> based on percentages across different service versions, or if you want to direct

Reviewer: this is either a run on or too complicated to understand imo.

Author: Yeah, it's a bit long. I'll see if I can rework.

traffic from your internal users to a particular set of instances.

With a virtual service, you can specify traffic behavior for one or more hostnames.
You use routing rules in the virtual service that tell Envoy how to send the
virtual service’s traffic to appropriate destinations. Route destinations can
be versions of the same service or entirely different services.

A typical use case is to send traffic to different versions of a service,

> specified as service subsets. Clients send requests to the virtual service host as if

Reviewer: subset needs a glossary definition.

Author: TODO

it was a single entity, and Envoy then routes the traffic to the different

> versions depending on the virtual service rules (for example "20% of calls go to

Reviewer: rework to remove parathesis.

Author: done.

the new version" or "calls from these users go to version 2"). This allows you to,

> for instance, create a canary rollout where you gradually increase the

Reviewer: not sure why you need the extra for instance in here.

Author: Creating a canary rollout is an example of what routing traffic to different versions depending on the rules allows you to do.

percentage of traffic that’s sent to a new service version. The traffic routing

> is completely separate from the instance deployment, meaning that the number of

Reviewer: deployment.  The number of...

Author: Though breaking it into two sentences loses the relationship between the clauses - the routing is separate from the deployment *which means that*/*because of which* the number of instances ....

instances implementing the new service version can scale up and down based on
traffic load without referring to traffic routing at all. By contrast, container
orchestration platforms like Kubernetes only support traffic distribution based
on instance scaling, which quickly becomes complex. You can read more about how

> virtual services help with canarying in [Canary Deployments using Istio](/blog/2017/0.1-canary/).

Reviewer: canaryign is not a verb :) .  `virtual services help with [Canary Deployments using Istio](link)`

Author: I have spent too long working with SREs. :)

Virtual services also let you:

-   Address multiple application services through a single virtual service. If
    your mesh uses Kubernetes, for example, you can configure a virtual service
    to handle all services in a specific namespace. Mapping a single

>     virtual service to multiple "real" services can be particularly useful in

Reviewer: can be -> is

Author: done.

    facilitating turning a monolithic application into a composite service built
    out of distinct microservices without requiring the consumers of the service

>     to adapt to the transition: your rules can specify "calls to these URIs of

Reviewer: change : to period.  Your rules which type of rules?

Author: done.

    `monolith.com` go to `microservice A`", and so on. You can see how this works
    in [one of our examples below](#more-about-routing-rules).
-   Configure traffic rules in combination with
    [gateways](/docs/concepts/traffic-management/#gateways) to control ingress
    and egress traffic.

In some cases you also need to configure destination rules to use these
features, as these are where you specify your service subsets. Specifying
service subsets and other destination-specific policies in a separate object
lets you reuse these cleanly between virtual services. You can find out more
about destination rules in the next section.

### Virtual service example {#virtual-service-example}

> Here’s a simple virtual service example. The following virtual service routes

Reviewer: You should be able to delete the first sentence without impairing the content.

Author: done.

    - /docs/concepts/traffic-management/pilot.html
---

-   [Introducing Istio Traffic Management](/docs/concepts/traffic-management/#introducing-istio-traffic-management):
    Learn about how Istio helps you control the flow of traffic in your mesh
    with simple, flexible configuration.

-   [Network resilience and testing](/docs/concepts/traffic-management/#network-resilience-and-testing):
    Learn about Istio's dynamic failure recovery features that you can configure
    to test and build tolerance for failing nodes, and to prevent cascading
    failures to other nodes.

-   [Architecture](/docs/concepts/traffic-management/#architecture): Learn about
    Pilot, Istio's core traffic management component and Envoy proxies and how

>     they enable service discovery and traffic control for services in the mesh.

Reviewer 3: I think this should be second

Author: Strongly disagree. In my original proposal I actually suggesting moving this content out of the doc entirely (into a separate architecture section for potential contributors & interested parties); though we agreed in the docs group meeting that we'd keep it at the end of this doc for now, largely because we haven't reorganized similar content out of any of the other concepts docs. In the current structure we're entirely focused on features and how they help the user until the arch. section at the end.

Reviewer 2: I think the 3 bullets at the top of the page obscure the fact that the document is feature focused, with architecture stuff moved to the back. Maybe the bullets aren't needed at all? Otherwise, I think it's fine for now, until we have a proper architecture section of the docs that we move the architecture section to.

Author: That's a good point too: I just reused the original bullets (with a bit of tweaking), I'm not entirely sure how much they help orient the reader beyond the TOC (which they can see on the right) and the actual intro.

Reviewer 2: I suggest removing them.

Author: Taken them out!

Istio’s traffic routing rules let you easily control the flow
of traffic and API calls between services. Istio simplifies configuration of
service-level properties like circuit breakers, timeouts, and retries, and makes
it easy to set up important tasks like A/B testing, canary rollouts, and staged
rollouts with percentage-based traffic splits. It also provides out-of-box
failure recovery features, which, combined with Istio’s
[observability](/docs/concepts/observability/) features, helps make your calls

> more reliable, and your network more robust.

Reviewer 3: helps make your application more robust against failures of dependent services or the network

you might want to direct a particular percentage of traffic to a new version of
a service as part of A/B testing, or apply a different load balancing policy to
traffic for a particular subset of service instances. You might also want to
apply special rules to traffic coming into or out of your mesh, or add an
external dependency of your mesh to the service registry. You can do all this
and more by adding your own traffic configuration to Istio using Istio’s traffic
management API.

Like other Istio configuration, the API is implemented as Kubernetes custom
resources (CRDs), which you can configure using YAML, as you’ll see in the
examples.

The rest of this guide looks at each of the traffic management API’s resources
and what you can do with them. These resources are:

-   [Virtual services](#virtual-services)
-   [Destination rules](#destination-rules)
-   [Gateways](#gateways)
-   [Service entries](#service-entries)
-   [Sidecars](#sidecars)

This guide also gives an overview of some of the
[network resilience and testing features](#network-resilience-and-testing) that
are built in to the API’s resources.

## Virtual services {#virtual-services}

> [Virtual services](/docs/reference/config/networking/v1alpha3/virtual-service/),

Reviewer 2: Maybe it would be better to reference the actual resources in the ref guide, rather than another overview.
```suggestion
[Virtual services](/docs/reference/config/networking/v1alpha3/virtual-service/#VirtualService),
```

> along with destination rules, are the key building blocks of Istio’s traffic

Reviewer 2: ```suggestion
along with [destination rules](/docs/reference/config/networking/v1alpha3/destination-rule/#DestinationRule), are the key building blocks of Istio’s traffic
```

    Learn about Istio's dynamic failure recovery features that you can configure
    to test and build tolerance for failing nodes, and to prevent cascading
    failures to other nodes.

-   [Architecture](/docs/concepts/traffic-management/#architecture): Learn about
    Pilot, Istio's core traffic management component and Envoy proxies and how
    they enable service discovery and traffic control for services in the mesh.

Istio’s rules configuration and traffic routing lets you easily control the flow
of traffic and API calls between services. Istio simplifies configuration of
service-level properties like circuit breakers, timeouts, and retries, and makes
it easy to set up important tasks like A/B testing, canary rollouts, and staged
rollouts with percentage-based traffic splits. It also provides out-of-box
failure recovery features, which, combined with Istio’s
[observability](/docs/concepts/observability/) features, helps make your calls
more reliable, and your network more robust.

Istio’s traffic management model relies on the Envoy proxies that are deployed
along with your services. All traffic that your mesh services send and receive

> (data plane traffic) is proxied through Envoy, making it easy to direct and

Author: Will do! Envoy proxies, anything else obvious?

> (data plane traffic) is proxied through Envoy, making it easy to direct and

Reviewer 2: "data plane" is another good choice for a glossary ref, but we don't have a glossary entry for it yet, so you'll need to add one (note that we already have "control plane" in the glossary).

> (data plane traffic) is proxied through Envoy, making it easy to direct and

Author: Ok, I'll put something in.

> (data plane traffic) is proxied through Envoy, making it easy to direct and

Reviewer 2: Other possible glossary refs: A/B testing, canary rollouts

> (data plane traffic) is proxied through Envoy, making it easy to direct and

Author: I've added data plane though I'm not sure how great it is (I think I last wrote about data planes in a networking course many years ago), PTAL.

> (data plane traffic) is proxied through Envoy, making it easy to direct and

Reviewer 2: LGTM

control traffic around your mesh without making any changes to your services -

> your workload just needs to specify the destination service, and the Envoy

Reviewer 2: Good point. Maybe just delete the whole `- your workload just ...` part, since it doesn't really say anything more than the previous sentence.

> your workload just needs to specify the destination service, and the Envoy

Author: That works. Done.

proxies do the rest.

If you’re interested in the details of how the features described in this guide
work, you can find out more about Istio’s traffic management architecture in the
[Architecture](#architecture) section at the end of this document. The rest of
this guide introduces Istio’s traffic management features.

## Introducing Istio Traffic Management

In order to direct traffic around your mesh, Istio needs to know where all your
endpoints are, and which services they belong to. To populate its own service
registry, Istio connects to a service discovery system. For example, if you've
installed Istio on Kubernetes, then by default Istio knows about all the

> services and endpoints in that cluster.

Reviewer 2: "that cluster" ?

Maybe the sentence should start with `For example, if you've installed Istio on a Kubernetes cluster ...`

Author: Better, thanks!

Reviewer 3: Add another line for Consul

load balancing pool. By default, the Envoy proxies distribute traffic across
each service’s load balancing pool using a round-robin model, where requests are
sent to each pool member in turn, returning to the top of the pool once each
service instance has received a request.

While this basic service discovery and load balancing gives you a working
service mesh, it’s far from all that Istio can do. In many cases you might want
more fine-grained control over what happens to your mesh traffic. For example,
you might want to direct a particular percentage of traffic to a new version of
a service as part of A/B testing, or apply a different load balancing policy to
traffic for a particular subset of service instances. You might also want to
apply special rules to traffic coming into or out of your mesh, or add an
external dependency of your mesh to the service registry. You can do all this
and more by adding your own traffic configuration to Istio using Istio’s traffic
management API.

Like other Istio configuration, the API is implemented as Kubernetes custom
resources (CRDs), which you can configure using YAML, as you’ll see in the

> examples.

Author: I'm hoping to do another concept doc on our config model/APIs, which could emphasise that even though you use the Kubernetes API server it doesn't mean the workloads etc. need to be on Kubernetes.

Is there a nice/short way of clarifying this here in the meantime without going off on a bit of a tangent from traffic management?

> examples.

Reviewer 2: Maybe the following wording would sound less tied to Kubernetes?
```
Like other Istio configuration, the API is specified using Kubernetes custom
resources definitions (CRDs), which you can configure using YAML, as you’ll see in the
examples.
```

> examples.

Reviewer 3: I don’t think we need to talk about crds here. This section should be generic.

> examples.

Author: I think it's useful context for the large chunks of YAML that we're about to show them - when I first started using Istio I was confused by the config/where exactly I was sending things to, then had a sort of lightbulb moment when I realized I was sending them to a Kubernetes API server.

> examples.

Reviewer 2: My first instinct was also to just not mention crds, but then as soon as you start looking at examples it's in your face, so I think it needs to say something.

> examples.

Reviewer 2: Just had a thought. Add a glossary reference for CRDs. In the Istio glossary have a short one liner with link to K8s doc about CRDs, but also say that Istio uses Kube CRD API for config, even for non-kube Istio deployments. WDYT?

> examples.

Author: Glossary entry is a good idea, as it means we don't need to shoehorn in any further explanation here/in other concept docs.

Would we still leave the sentence as per your suggestion and then link to the new definition on "CRDs"?

> examples.

Reviewer 2: That works for me

> examples.

Author: Done!

The rest of this guide looks at each of the traffic management API’s resources
and what you can do with them. These resources are:

-   [Virtual services](#virtual-services)
-   [Destination rules](#destination-rules)
-   [Gateways](#gateways)

> -   [Service Entries](#service-entries)

Reviewer 4: Lowecase e: entries

Author: done!

-   [Sidecars](#sidecars)

This guide also gives an overview of some of the
[network resilience and testing features](#network-resilience-and-testing) that
are built in to the API’s resources.

## Virtual services {#virtual-services}

[Virtual services](/docs/reference/config/networking/v1alpha3/virtual-service/),
along with destination rules, are the key building blocks of Istio’s traffic
routing functionality. A virtual service lets you configure how requests are
routed to a service within an Istio service mesh, building on the basic
connectivity and discovery provided by Istio and your platform. Each virtual
service consists of a set of routing rules that are evaluated in order, letting
Istio match each given request to the virtual service to a specific real
destination within the mesh. Your mesh can require multiple virtual services or
none depending on your use case.

### Why use virtual services? {#why-use-virtual-services}

Virtual services play a key role in making Istio’s traffic management flexible

> and powerful. They do this by strongly decoupling where the client sends their

Reviewer 2: ```suggestion
and powerful. They do this by strongly decoupling where clients send their
```

requests from the destination workloads that actually implement them. Virtual
services also provide a rich way of specifying different traffic routing rules
for sending traffic to those workloads.

So why is this so useful? Without virtual services, Envoy just distributes
traffic using round-robin load balancing between all service instances, as
described in the introduction. You can improve this behavior with what you know
about the workloads: for example, some might represent a different version. This
can be useful in A/B testing where you might want to configure traffic routes

> based on percentages across different service versions, or you might want to use

Reviewer 2: You mention "percentage-based" routing several time, but I think it would be good to be clear that you can distribute request based on other criteria as well (e.g., internal users vs. everyone else).

Author: Have updated (and also removed the poor example from your previous comment...).

> a different load balancing policy for a particular set of instances.

Reviewer 2: load balancing policy is configured in destination rules, not virtual services.

Author: Well-spotted, it's not a great example. Will come up with something else....

Author: Done.

> With virtual services, you create a new hostname which directs traffic to

Reviewer 5: > you create a new hostname

This reads like we are creating DNS entries, which we aren't

Reviewer 2: I agree. How about changing this whole paragraph to something like this:
```
With a virtual service, you can specify traffic behavior for one or more hostnames.
You use routing rules in the virtual service that tell Envoy how to send the
virtual service’s traffic to appropriate destinations. Route destinations can
be versions of the same service or a different service.
```

Author: Nice. Done.

services of your choice, or overrides traffic behavior for an existing hostname.
You can then use routing rules in the virtual service that tell Envoy how to

> send the virtual host’s traffic to the appropriate "real" workloads.

Reviewer 5: ```suggestion
send the virtual host’s traffic to the appropriate destination.
```

A typical use case is to send traffic to different versions of a service,

> specified as service subsets. Clients send requests to the virtual service as if

Reviewer 5: Do we want to say the send requests "to the virtual service"? Or they send requests to the service and the virtual service routes it?

Reviewer 2: Maybe `Clients send requests to the virtual service host as if ...`?

Author: That works. Done.

it was a single entity, and Envoy then routes the traffic to the different
versions depending on the virtual service rules (for example "20% of calls go to

> the new version" or "calls from this URI go to version 2"). This allows you to,

Reviewer 2: ```suggestion
the new version" or "calls to this URI go to version 2"). This allows you to,
```
A more realistic example would be something like:
```
"calls to these URIs of some host (e.g., `monolith.com`) go to `microserviceA`"
```

Author: So I've changed the example here and expanded the bit below about directing traffic based on URI, see if you think that works.

Reviewer 2: Very good!

for instance, create a canary rollout where you gradually increase the
percentage of traffic that’s sent to a new service version. The traffic routing
is completely separate from the instance deployment, meaning that the number of
instances implementing the new service version can scale up and down based on
traffic load without referring to traffic routing at all. By contrast, container

> orchestration platforms like Docker or Kubernetes support canary rollouts, but

Reviewer 5: suggest:

 By contrast, container orchestration platforms like Kubernetes only support traffic distribution based on instance scaling, which quickly becomes complex.

Docker is not really an orchestration platform, and I wouldn't consider K8s supporting canary rollouts -- and saying they do takes away from the Istio value add

Author: Oh I like that. Will change.

Reviewer 2: The current doc has a link to [my old blog post](https://archive.istio.io/v1.1/blog/2017/0.1-canary/) that better explains the difference with orchestration platform support. We added the link from the concepts because it was the only doc we had on the subject and because it's an old blog post users were having trouble finding it. I think we should still have a link unless we move the material from the blog into the docs somewhere that is easier to find.

Author: I'll put the link back in, that content is very useful. I'll have a think about where we might put it in the main docset.

they use instance scaling to manage traffic distribution, which quickly becomes
complex.

Virtual services also let you:

-   Address multiple application services through a single virtual service. If
    your mesh uses Kubernetes, for example, you can configure a virtual service
    to handle all services in a specific namespace, as you can see in
    [one of our examples below](#more-about-routing-rules). Mapping a single
    virtual service to multiple "real" services can be particularly useful in
    facilitating turning a monolithic application into a composite service built
    out of distinct microservices without requiring the consumers of the service
    to adapt to the transition.
-   Configure traffic rules in combination with
    [gateways](/docs/concepts/traffic-management/#gateways) to control ingress
    and egress traffic.

> In most cases you also need to configure destination rules to use these

Reviewer 5: ```suggestion
In some cases you also need to configure destination rules to use these
```

  hosts:
  - reviews
  http:
  - match:
    - headers:
        end-user:
          exact: jason
    route:
    - destination:
        host: reviews
        subset: v2
  - route:
    - destination:
        host: reviews
        subset: v3
{{< /text >}}

#### The hosts field {#the-hosts-field}

The `hosts` field lists the virtual service’s hosts - in other words, the user-addressable
destination or destinations that these routing rules apply to. This is the
address or addresses the client uses when sending requests to the service.

{{< text yaml >}}
hosts:
- reviews
{{< /text >}}

The virtual service hostname can be an IP address, a DNS name, or, depending on
the platform, a short name (such as a Kubernetes service short name) that resolves,
implicitly or explicitly, to a fully qualified domain name (FQDN). You can also
use wildcard ("\*") prefixes, letting you create a single set of routing rules for
all matching services. Virtual service hosts don't actually have to be part of the
Istio service registry, they are simply virtual destinations. This lets you model
traffic for virtual hosts that don't have routable entries inside the mesh.

> #### Routing rules {#routing-rules}

Reviewer: note to self - review from here forward.

  hosts:
  - reviews
  http:
  - match:
    - headers:
        end-user:
          exact: jason
    route:
    - destination:
        host: reviews
        subset: v2
  - route:
    - destination:
        host: reviews
        subset: v3
{{< /text >}}

#### The hosts field {#the-hosts-field}

> `hosts` lists the virtual service’s hosts - in other words, the user-addressable

Reviewer 2: ```suggestion
The `hosts` field lists the virtual service’s hosts - in other words, the user-addressable
```

use wildcard ("\*") prefixes, letting you create a single set of routing rules for
all matching services. Virtual service hosts don't actually have to be part of the
Istio service registry, they are simply virtual destinations. This lets you model
traffic for virtual hosts that don't have routable entries inside the mesh.

#### Routing rules {#routing-rules}

The `http` section contains the virtual service’s routing rules, describing
match conditions and actions for routing HTTP/1.1, HTTP2, and gRPC traffic sent
to the destination(s) specified in the hosts field (you can also use `tcp` and
`tls` sections to configure routing rules for
[TCP](/docs/reference/config/networking/v1alpha3/virtual-service/#TCPRoute) and
unterminated
[TLS](/docs/reference/config/networking/v1alpha3/virtual-service/#TLSRoute)
traffic). A routing rule consists of the destination where you want the traffic
to go and zero or more match conditions, depending on your use case.

##### Match condition {#match-condition}

The first routing rule in the example has a condition and so begins with the
`match` field. In this case you want this routing to apply to all requests from

> the user "jason", so you use the headers, end-user, and exact fields to select

Reviewer 2: ```suggestion
the user "jason", so you use the `headers`, `end-user`, and `exact` fields to select
```

the appropriate requests.

{{< text yaml >}}
- match:
   - headers:
       end-user:
         exact: jason
{{< /text >}}

##### Destination {#destination}

> The route section’s destination field specifies the actual destination for

Reviewer 2: ```suggestion
The route section’s `destination` field specifies the actual destination for
```

traffic that matches this condition. Unlike the virtual service’s host(s), the
destination’s host must be a real destination that exists in Istio’s service
registry or Envoy won’t know where to send traffic to it. This can be a mesh

> service with proxies or a non-mesh service added using a Service Entry. In this

Reviewer 2: ```suggestion
service with proxies or a non-mesh service added using a service entry. In this
```

namespace. Because using the Kubernetes short name can result in
misconfigurations, we recommend that you specify fully qualified host names in
production environments.
{{< /tip >}}

The destination section also specifies which subset of this Kubernetes service
you want requests that match this rule’s conditions to go to, in this case the
subset named v2. You’ll see how you define a service subset in the section on
[destination rules](#destination-rules) below.

#### Routing rule precedence {#routing-rule-precedence}

Routing rules are **evaluated in sequential order from top to bottom**, with the
first rule in the virtual service definition being given highest priority. In
this case you want anything that doesn't match the first routing rule to go to a
default destination, specified in the second rule. Because of this, the second
rule has no match conditions and just directs traffic to the v3 subset.

{{< text yaml >}}
- route:
  - destination:
      host: reviews
      subset: v3
{{< /text >}}

We recommend providing a default "no condition" or weight-based rule (described
below) like this as the last rule in each virtual service to ensure that traffic

> to the virtual service always has at least one matching destination.

Reviewer 2: ```suggestion
to the virtual service always has at least one matching route.
```

### More about routing rules {#more-about-routing-rules}

As you saw above, routing rules are a powerful tool for routing particular
subsets of traffic to particular destinations. You can set match conditions on
traffic ports, header fields, URIs, and more. For example, this virtual service
lets users send traffic to two separate services, ratings and reviews, as if
they were part of a bigger virtual service at `http://bookinfo.com/.` The
virtual service rules match traffic based on request URIs and direct requests to
the appropriate service.

{{< text yaml >}}
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: bookinfo
spec:
  hosts:
    - bookinfo.com
  http:
  - match:
    - uri:
        prefix: /reviews
    route:
    - destination:
        host: reviews
  - match:
    - uri:
        prefix: /ratings
    route:
    - destination:
        host: ratings
...

  http:
  - match:
      sourceLabels:
        app: reviews
    route:
...
{{< /text >}}

For some match conditions, you can also choose to select them using the exact
value, a prefix, or a regex.

You can add multiple match conditions to the same `match` block to AND your

> conditions, or add multiple match blocks to the same rule to OR your conditions.

Reviewer 2: Hmm. We used to have AND and OR match examples here, which people found helpful because I believe they are the only such examples we have anywhere in the docs.

Author: Hmm indeed. I was trying to cut down the number of examples because this is meant to be an introductory concepts doc, not a config reference (if your concepts docs/tutorials have become a reference it's generally a sign that content needs to be moved....). I thought everything was covered somewhere else but apparently not. Is there another good place to put the AND and OR examples?

Reviewer 2: I guess they should go into the ref docs. Even without examples, with your text in the concepts that says AND and OR is possible, people will hopefully be able to figure it out. Let's go with that assumption for now.

Author: OK. Should we add an issue to add them to the refdocs?

Reviewer 2: Let's wait to see if they're really needed.

You can also have multiple routing rules for any given virtual service. This
lets you make your routing conditions as complex or simple as you like within a
single virtual service. A full list of match condition fields and their possible
values can be found in the

> [Virtual Service reference](/docs/reference/config/networking/v1alpha3/virtual-service/#HTTPMatchRequest).

Reviewer 2: ```suggestion
[HTTPMatchRequest reference](/docs/reference/config/networking/v1alpha3/virtual-service/#HTTPMatchRequest).
```

In addition to using match conditions, you can distribute traffic that
distributes a particular routing by percentage "weight", useful for A/B testing
and canary rollouts:

{{< text yaml >}}
spec:
  hosts:
  - reviews
  http:
  - route:
    - destination:
        host: reviews
        subset: v1
      weight: 75
    - destination:
        host: reviews
        subset: v2
      weight: 25
{{< /text >}}

You can also use routing rules to perform some actions on the traffic, for
example:

-   Append or remove headers.
-   Rewrite the URL.
-   Set a [retry policy](#retries) for calls to this destination.

To learn more about the actions available, see the

> [Virtual Service reference](/docs/reference/config/networking/v1alpha3/virtual-service/#HTTPRoute).

Reviewer 2: ```suggestion
[HTTPRoute reference](/docs/reference/config/networking/v1alpha3/virtual-service/#HTTPRoute).
```

## Destination rules {#destination-rules}

Along with [virtual services](#virtual-services),

> [destination rules](/docs/reference/config/networking/v1alpha3/destination-rule/)

Reviewer 2: ```suggestion
[destination rules](/docs/reference/config/networking/v1alpha3/destination-rule/#DestinationRule)
```

are a key part of Istio’s traffic routing functionality. You can think of
virtual services as how you route your traffic **to** a given destination, and
then you use destination rules to configure what happens to traffic **for** that
destination. Destination rules are applied after virtual service routing rules
are evaluated, so they apply to the traffic’s "real" destination.

In particular, you use destination rules to specify named service subsets, such
as grouping all a given service’s instances by version. Defining subsets in

> destination rules lets you cleanly refer to a specific service version across

Reviewer 2: This sentence seems repetitive. I read the exact same thing in the previous section.

Author: You did indeed because I think I accidentally copied it rather than cut and pasted it. :)

  hosts:
  - reviews
  http:
  - match:
    - headers:
        end-user:
          exact: jason
    route:
    - destination:
        host: reviews
        subset: v2
  - route:
    - destination:
        host: reviews
        subset: v3
{{< /text >}}

#### The hosts field {#the-hosts-field}

`hosts` lists the virtual service’s hosts - in other words, the user-addressable
destination or destinations that these routing rules apply to. This is the
address or addresses the client uses when sending requests to the service.

{{< text yaml >}}
hosts:
- reviews
{{< /text >}}

> The virtual service hostname can be a DNS name or an internal mesh service name

Reviewer 5: I don't really understand this sentence, so others may find it confusing as well.

Author: Let me see if I can clarify it a bit. PTAL!

as long as the name resolves, implicitly or explicitly, to one or more fully

> qualified domain names (FQDN). You can also use wildcard ("*") prefixes, letting

Reviewer 5: Github does not  like the `*` -- not sure if this is just github or we need to escape it

Author: Think it might be just Github as it renders in the preview, I tried escaping it and the output was identical...

the appropriate requests.

{{< text yaml >}}
- match:
   - headers:
       end-user:
         exact: jason
{{< /text >}}

##### Destination {#destination}

The route section’s destination field specifies the actual destination for
traffic that matches this condition. Unlike the virtual service’s host(s), the
destination’s host must be a real destination that exists in Istio’s service
registry or Envoy won’t know where to send traffic to it. This can be a mesh
service with proxies or a non-mesh service added using a Service Entry. In this
case we’re running on Kubernetes and the host name is a Kubernetes service name:

{{< text yaml >}}
route:
- destination:
    host: reviews
    subset: v2
{{< /text >}}

{{< tip >}}
In this and other examples, we use a Kubernetes short name for the
destination hosts. When this rule is evaluated, Istio adds a domain suffix based
on the namespace of the virtual service that contains the routing rule to get
the fully qualified name for the host. However, this only works if the
destination hosts and the virtual service are actually in the same Kubernetes
namespace. Because using the Kubernetes short name can result in

> misconfigurations, we recommend that you specify fully qualified host names in

Reviewer 5: meta point: if we recommend it shouldn't we use that in documentation

Reviewer 2: The main reason our samples/examples use short names is so that they are namespace-independent, so users can deploy them in any namespace they'd like. That said, maybe the examples in this concepts section could use long names since, unlike tasks, the user isn't expected to try them. Either that, or maybe we should make this tip a little more clear about why we use short names here (e.g., for simplicity) and in the tasks (namespace independence convenience).

Author: I think if we add "...for simplicity" in the tip to clarify it should make it clear that we're just doing it so the examples are very easy to read.

Author: Done, PTAL once I've committed everything...

Reviewer 2: That's perfect.

Just a thought: would it be better to make the first part of the tip just an ordinary paragraph, and make the second part a warning instead of a tip? Something like this:
```
In this and the other examples on this page, we use a Kubernetes short name for the
destination hosts for simplicity. When this rule is evaluated, Istio adds a domain suffix based
on the namespace of the virtual service that contains the routing rule to get
the fully qualified name for the host. Using short names in our examples
also means that you can copy and try them in any namespace you like.

{{< warning >}}
Using short names like this only works if the
destination hosts and the virtual service are actually in the same Kubernetes
namespace. Because using the Kubernetes short name can result in
misconfigurations, we recommend that you specify fully qualified host names in
production environments.
{{< /warning >}}
```

Author: Oh yeah that works, and means we don't have a giant tip box.

      version: v3
{{< /text >}}

Each subset is defined based on one or more `labels`, which in Kubernetes are
key/value pairs that are attached to objects such as Pods. These labels are
applied in the Kubernetes service’s deployment as `metadata` to identify
different versions.

As well as defining subsets, this destination rule has both a default traffic
policy for all subsets in this destination and a subset-specific policy that
overrides it for just that subset. The default policy, defined above the subsets
field, sets a simple random load balancer for the `v1` and `v3` subsets. In the
`v2` policy, a round-robin load balancer is specified in the corresponding
subset’s field.

## Gateways {#gateways}

> You use a [gateway](/docs/reference/config/networking/v1alpha3/gateway/) to

Reviewer 2: ```suggestion
You use a [gateway](/docs/reference/config/networking/v1alpha3/gateway/#Gateway) to
```

> overrides it for just that subset. The default policy, defined above the subsets

Reviewer 2: ```suggestion
overrides it for just that subset. The default policy, defined above the `subsets`
```

{{< /text >}}

> This gateway configuration lets HTTPS traffic from `ext-host` into the mesh on

Reviewer 2: `ext-host` is a silly example of an external host name. Maybe change to something like `ext-host.example.com`?

Author: Good spot. Fixed.

port 443, but doesn’t specify any routing for the traffic.

To specify routing and for the gateway to work as intended, you must also bind
the gateway to a virtual service. You do this using the virtual service’s
`gateways` field, as shown in the following example:

> {{< text yaml >}}

Reviewer 2: The host in the virtual service needs to match the gateway host ext-host (or ext-host.example.com), not ext-svc

Author: Ditto!

You can then configure the virtual service with routing rules for the external
traffic.

## Service entries {#service-entries}

You use a

> [service entry](/docs/reference/config/networking/v1alpha3/service-entry) to add

Reviewer 2: ```suggestion
[service entry](/docs/reference/config/networking/v1alpha3/service-entry/#ServiceEntry) to add
```

was a service in your mesh. Configuring service entries allows you to manage
traffic for services running outside of the mesh, including the following tasks:

-   Redirect and forward traffic for external destinations, such as APIs
    consumed from the web, or traffic to services in legacy infrastructure.
-   Define [retry](#retries), [timeout](#timeouts), and
    [fault injection](#fault-injection) policies for external destinations.
-   Add a service running in a Virtual Machine (VM) to the mesh to
    [expand your mesh](/docs/examples/mesh-expansion/single-network/#running-services-on-a-mesh-expansion-machine).
-   Logically add services from a different cluster to the mesh to configure a
    [multicluster Istio mesh](/docs/setup/install/multicluster/gateways/#configure-the-example-services)
    on Kubernetes.

You don’t need to add a service entry for every external service that you want
your mesh services to use. By default, Istio configures the Envoy proxies to
passthrough requests to unknown services. However, you can’t use Istio features
to control the traffic to destinations that aren't registered in the mesh.

### Service entry example {#service-entry-example}

> The following example mesh-external service entry adds the `ext-resource`

Reviewer 2: ```suggestion
The following example mesh-external service entry adds the `ext-resource.com`
```

Reviewer 2: maybe a better name would be `ext-svc.example.com`?

Author: Agreed. Done.

  resolution: DNS
{{< /text >}}

> You specify the external resource using the `hosts` key. You can qualify it

Reviewer 2: ```suggestion
You specify the external resource using the `hosts` field. You can qualify it
```

> fully or use a wildcard domain name.

Reviewer 2: ```suggestion
fully or use a wildcard prefixed domain name.
```

You can configure virtual services and destination rules to control traffic to a
service entry in a more granular way, in the same way you configure traffic for
any other service in the mesh. For example, the following destination rule
configures the traffic route to use mutual TLS to secure the connection to the

> `ext-resource` external service that we configured using the service entry:

Reviewer 2: ```suggestion
`ext-resource.com` external service that we configured using the service entry:
```

      caCertificates: /etc/certs/rootcacerts.pem
{{< /text >}}

See the
[Service Entry reference](/docs/reference/config/networking/v1alpha3/service-entry)
for more possible configuration options.

## Sidecars {#sidecars}

By default, Istio configures every Envoy proxy to accept traffic on all the
ports of its associated workload, and to reach every workload in the mesh when

> forwarding traffic. You can use a sidecar configuration to do the following:

Reviewer 2: ```suggestion
forwarding traffic. You can use a
[sidecar](/docs/reference/config/networking/v1alpha3/sidecar/#Sidecar)
configuration to do the following:
```

for more details.

## Network resilience and testing {#network-resilience-and-testing}

As well as helping you direct traffic around your mesh, Istio provides opt-in
failure recovery and fault injection features that you can configure dynamically
at runtime. Using these features helps your applications operate reliably,
ensuring that the service mesh can tolerate failing nodes and preventing
localized failures from cascading to other nodes.

### Timeouts {#timeouts}

A timeout is the amount of time that an Envoy proxy should wait for replies from
a given service, ensuring that services don’t hang around waiting for replies
indefinitely and that calls succeed or fail within a predictable timeframe. The
default timeout for HTTP requests is 15 seconds, which means that if the service
doesn’t respond within 15 seconds, the call fails.

For some applications and services, Istio’s default timeout might not be
appropriate. For example, a timeout that is too long could result in excessive
latency from waiting for replies from failing services, while a timeout that is
too short could result in calls failing unnecessarily while waiting for an

> operation involving multiple services to return. To find your optimal timeout

Reviewer 2: "To find your optimal ..."? Do you mean "To tune your optimal ..."? Or maybe "To find and use your optimal ..."

Author: The latter!

After you’ve configured your network, including failure recovery policies, you
can use Istio’s fault injection mechanisms to test the failure recovery capacity
of your application as a whole. Fault injection is a testing method that
introduces errors into a system to ensure that it can withstand and recover from
error conditions. Using fault injection can be particularly useful to ensure
that your failure recovery policies aren’t incompatible or too restrictive,
potentially resulting in critical services being unavailable.

Unlike other mechanisms for introducing errors such as delaying packets or
killing pods at the network layer, Istio’ lets you inject faults at the
application layer. This lets you inject more relevant failures, such as HTTP
error codes, to get more relevant results.

You can inject two types of faults, both configured using a
[virtual service](#virtual-services):

-   Delays: Delays are timing failures. They mimic increased network latency or
    an overloaded upstream service.
-   Aborts: Aborts are crash failures. They mimic failures in upstream services.
    Aborts usually manifest in the form of HTTP error codes or TCP connection
    failures.

> For example, this virtual service introduces a 5 second delay for 10% of the

Reviewer 2: ```suggestion
For example, this virtual service introduces a 5 second delay in 1 out of every 1000
```

        subset: v1
{{< /text >}}

For detailed instructions on how to configure delays and aborts, see
[Fault Injection](/docs/tasks/traffic-management/fault-injection/).

### Working with your applications {#working-with-your-applications}

> -   Istio failure recovery features are completely transparent to the

Reviewer 2: One more nit: why are these bullets instead of paragraphs?

Author: Good point. I think they were small bullets when I started and then grew....

I'll de-bullet them.

Istio's traffic management model relies on the following two components:

-   {{< gloss >}}Pilot{{</ gloss >}}, the core traffic management component.
-   {{< gloss >}}Envoy{{</ gloss >}} proxies, which enforce configurations and
    policies set through Pilot.

These components enable the following Istio traffic management features:

-   Service discovery
-   Load balancing
-   Traffic routing and control

### Pilot: Core traffic management {#pilot}

The following diagram shows the Pilot architecture:

{{< image width="40%" link="./pilot-arch.svg" caption="Pilot architecture" >}}

As the diagram illustrates, Pilot maintains an **abstract model** of all the
services in the mesh. **Platform-specific adapters** in Pilot translate the
abstract model appropriately for your platform. For example, the Kubernetes
adapter implements controllers to watch the Kubernetes API server for changes to
pod registration information and service resources. The Kubernetes adapter
translates this data for the abstract model.

Pilot uses the abstract model to generate appropriate Envoy-specific
configurations to let Envoy proxies know about one another in the mesh through

> the **Envoy API.**

Reviewer 2: ```suggestion
the [Envoy API](https://www.envoyproxy.io/docs/envoy/latest/api/api).
```

> You can use Istio's **Traffic Management API** to instruct Pilot to refine the

Reviewer 2: ```suggestion
You can use Istio's [Traffic Management API](#traffic-routing-and-configuration) to instruct Pilot to refine the
```

Envoy configuration to exercise more granular control over the traffic in your
service mesh.

### Envoy proxies

Traffic in Istio is categorized as data plane traffic and control plane traffic.

> Data plane traffic refers to the data that the business logic of the workloads

Reviewer 2: ```suggestion
Data plane traffic refers to the messages that the business logic of the workloads
```

> manipulate. Control plane traffic refers to configuration and control data sent

Reviewer 2: ```suggestion
send and receive. Control plane traffic refers to configuration and control messages sent
```
